### PR TITLE
Visual Script Refactor

### PR DESCRIPTION
**Pull request by swarnimarun**
Edited: https://github.com/swarnimarun/godot/pull/7 Depends on this.

Refactor of visual script internals and dependent editor code, for simplifying further code changes for GSOC.

Things left to do:
- [x] Discuss some functionality changes. (I will add images and explanation for them ASAP)
- [x] Finish some leftover TODOs based on these functionality changes.
- [x] Squash commits when everything is done.
- [x] Test the code further.

cc: @fire @kidrigger

### CODE REVIEW

**Comment by fire** [comment]
Can you write things I can do to test this?

**Comment by swarnimarun** [comment]
@fire feel free to build and try, https://github.com/swarnimarun/godot/tree/visualscript-refactor-04062020

Few things that won't work atm, the data connections aren't working atm internally so trying to get the code to run wouldn't work properly. 

This is related to the discussion I asked for above, the problem is relatively straight forward, whether or not to handle data input connections with nodes from out of the sequence tree, inside visual script instance, or in the editor at the time of editing.

image
Basically as seen above three different functions go to the same node but it has a data connection from only one, this could be an error at runtime or just use the default value for the port in the other functions, or it can be an error in the editor itself and just disconnect connections when it's unallowed behaviour and not let the internals care about any of it.

I am leaning towards making it a hard error in the editor itself.

As for what should happen right now, it basically just uses the default value of the data input port.
So the above prints lots of `Null`'s
image

Also, there are some things that can possibly be optimized but would lead to fragile interfaces, such as `get_available_id` should always return the current maximum valued id. Which can be emulated by just increasing a class instance variable counter by one for each count to `get_available_id` function. But that would mean every call to the function will give a different output rather than just give the maximum id.
I have made sure to keep things like these away as per our motives of keeping the codebase simple.

**Comment by swarnimarun** [comment]
I tested it a bit more and it should be ready for reviews and merge.
I will squash it if nothing more seems necessary.

cc: @fire @kidrigger

**Comment by fire** [comment]
Can you make the pull request on Godot Engine org? so it's easier to merge? Also squash and there should be a proposal.

**Comment by fire** [comment]
Reminder to squash. Thank you.

**Comment by swarnimarun** [comment]
@fire done. 👍 
WeW force pushing makes me more nervous than Horror movies... LMAO

**Comment by fire** [comment]
How do I test this? What features are being added?

**Comment by swarnimarun** [comment]
> How do I test this? What features are being added?

Everything that worked before should work.
I will try to re-write the demo projects, with this PR and throw it here this weekend I guess that should be enough to fiddle around with.

**Comment by fire** [comment]
> This branch has conflicts that must be resolved

**Comment by fire** [comment]
I rebased the branch on master.

File: modules/visual_script/visual_script.cpp [hunk 1]
```diff
@@ -55,9 +55,8 @@ void VisualScriptNode::set_default_input_value(int p_port, const Variant &p_valu
 	default_input_values[p_port] = p_value;
 
 #ifdef TOOLS_ENABLED
-	for (Set<VisualScript *>::Element *E = scripts_used.front(); E; E = E->next()) {
-		E->get()->set_edited(true);
-	}
+	if (script_used.is_valid())
+		script_used->set_edited(true);
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:59]:
```suggestion
	if (script_used.is_valid()) {
		script_used->set_edited(true);
	}
```

File: modules/visual_script/visual_script.cpp [hunk 2]
```diff
@@ -165,13 +162,15 @@ VisualScriptNodeInstance::~VisualScriptNodeInstance() {
 	}
 }
 
-void VisualScript::add_function(const StringName &p_name) {
+void VisualScript::add_function(const StringName &p_name, int func_node_id) {
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:165]:
```suggestion
void VisualScript::add_function(const StringName &p_name, int p_func_node_id) {
```

File: modules/visual_script/visual_script.cpp [hunk 3]
```diff
@@ -165,13 +162,15 @@ VisualScriptNodeInstance::~VisualScriptNodeInstance() {
...
 	ERR_FAIL_COND(instances.size());
 	ERR_FAIL_COND(!String(p_name).is_valid_identifier());
 	ERR_FAIL_COND(functions.has(p_name));
+	ERR_FAIL_COND(variables.has(p_name)); // TODO: Check if this would even cause a clash
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:169]: Did you get to check this TODO?

File: modules/visual_script/visual_script.cpp [hunk 4]
```diff
@@ -182,11 +181,7 @@ void VisualScript::remove_function(const StringName &p_name) {
 	ERR_FAIL_COND(instances.size());
 	ERR_FAIL_COND(!functions.has(p_name));
 
-	for (Map<int, Function::NodeData>::Element *E = functions[p_name].nodes.front(); E; E = E->next()) {
-		E->get().node->disconnect("ports_changed", callable_mp(this, &VisualScript::_node_ports_changed));
-		E->get().node->scripts_used.erase(this);
-	}
-
+	// let the editor handle the node removal
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:184]:
```suggestion
	// Let the editor handle the node removal.
```

File: modules/visual_script/visual_script.cpp [hunk 5]
```diff
@@ -281,259 +259,192 @@ void VisualScript::_node_ports_changed(int p_id) {
...
-	for (Map<StringName, Function>::Element *E = functions.front(); E; E = E->next()) {
-		ERR_FAIL_COND(E->get().nodes.has(p_id)); //id can exist only one in script, even for different functions
-	}
-
-	Function &func = functions[p_func];
-
-	if (Object::cast_to<VisualScriptFunction>(*p_node)) {
-		//the function indeed
-		ERR_FAIL_COND_MSG(func.function_id >= 0, "A function node has already been set here.");
+	ERR_FAIL_COND(nodes.has(p_id)); //id can exist only one in script
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:275]:
```suggestion
	ERR_FAIL_COND(nodes.has(p_id)); // ID can exist only one in script.
```

File: modules/visual_script/visual_script.cpp [hunk 6]
```diff
@@ -674,30 +582,25 @@ void VisualScript::rename_variable(const StringName &p_name, const StringName &p
 
 	ERR_FAIL_COND(!String(p_new_name).is_valid_identifier());
 
-	ERR_FAIL_COND(functions.has(p_new_name));
+	ERR_FAIL_COND(functions.has(p_new_name)); //why?
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:585]: Good question, which needs to be resolved :)

File: modules/visual_script/visual_script.cpp [hunk 7]
```diff
@@ -808,17 +711,16 @@ void VisualScript::get_custom_signal_list(List<StringName> *r_custom_signals) co
 }
 
 int VisualScript::get_available_id() const {
-	int max_id = 0;
-	for (Map<StringName, Function>::Element *E = functions.front(); E; E = E->next()) {
-		if (E->get().nodes.empty()) {
-			continue;
+	// TODO: Maybe add a class instance counter here to get the available id??
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:714]: What's the status here?

File: modules/visual_script/visual_script.cpp [hunk 8]
```diff
@@ -875,17 +779,18 @@ ScriptInstance *VisualScript::instance_create(Object *p_this) {
...
 		Map<StringName, Variant> values;
 
-		for (Map<StringName, Variable>::Element *E = variables.front(); E; E = E->next()) {
-			if (!E->get()._export) {
+		List<StringName> keys;
+		variables.get_key_list(&keys);
+
+		for (const List<StringName>::Element *E = keys.front(); E; E = E->next()) {
+			if (!variables[E->get()]._export)
 				continue;
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:787]:
```suggestion
			if (!variables[E->get()]._export) {
				continue;
			}
```

File: modules/visual_script/visual_script.cpp [hunk 9]
```diff
@@ -988,15 +896,14 @@ bool VisualScript::has_method(const StringName &p_method) const {
 }
 
 MethodInfo VisualScript::get_method_info(const StringName &p_method) const {
-	const Map<StringName, Function>::Element *E = functions.find(p_method);
-	if (!E) {
+	const Function funct = functions[p_method];
+	if (funct.func_id == -1)
 		return MethodInfo();
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:901]:
```suggestion
	if (funct.func_id == -1) {
		return MethodInfo();
	}
```

File: modules/visual_script/visual_script.cpp [hunk 10]
```diff
@@ -1430,20 +1280,20 @@ bool VisualScriptInstance::get(const StringName &p_name, Variant &r_ret) const {
 }
 
 void VisualScriptInstance::get_property_list(List<PropertyInfo> *p_properties) const {
-	for (const Map<StringName, VisualScript::Variable>::Element *E = script->variables.front(); E; E = E->next()) {
-		if (!E->get()._export) {
+	List<StringName> vars;
+	script->variables.get_key_list(&vars);
+	for (const List<StringName>::Element *E = vars.front(); E; E = E->next()) {
+		if (!script->variables[E->get()]._export)
 			continue;
```
Comment (akien-mga) [inline modules/visual_script/visual_script.cpp:1287]:
```suggestion
		if (!script->variables[E->get()]._export) {
			continue;
		}
```
There might be more, make sure that all single-line blocks use brackets as required by our style guide now.

File: modules/visual_script/visual_script.cpp (hunk 7 above)
Comment (swarnimarun) [reply modules/visual_script/visual_script.cpp:714]: Oh I decided against that as that seemed wasteful. Will remove all the useless TODO comments. :)

File: modules/visual_script/visual_script.cpp (hunk 10 above)
Comment (swarnimarun) [reply modules/visual_script/visual_script.cpp:1287]: Ahh thanks for it, almost forgot, will review and fix up all of them.

Comment (fire) [reply modules/visual_script/visual_script.cpp:1287]: Ping!

**Comment by jak6jak** [comment]
So, I was testing this pull request out and it seems that it causes Godot to crash on start up. The crash happens in VisualScript::get_script_method_list because the hash map nodes is empty and it tries to retrieve elements from it. I'll continue trying to debug this and hopefully find a fix. @swarnimarun do you know where nodes should be populated?

**Comment by swarnimarun** [comment]
@jak6jak Thanks a lot for the testing, I had been fairly busy with college work and exams, but I am fairly certain that it shouldn't be happening, at least it didn't use to happen back when I had tested it while I was proactively working on it. Maybe I had broken something I didn't realize... :P 
Will have to recheck some of the things.

Looking into it, let's see what more needs fixing...

**Comment by jak6jak** [comment]
Hey, @swarnimarun is this commit not supposed to be compatible with older projects?  I think the issue is related to opening projects before the patch. Here is a project I made using master to test on (I made it a .git repo for easy resetting in case files are being corrupted) : [test.zip](https://github.com/godotengine/godot/files/5351391/test.zip)
When you open it on the refactor it fails to load the nodes in the graph and adds a random function that shouldn't exist. Opening the scene crashes it because of the above issue.
Don't worry about being too busy I can help you finish things up.

**Comment by swarnimarun** [comment]
@jak6jak Yep it's not gonna be able to maintain compat.

**Comment by swarnimarun** [comment]
@fire Mind adding a breaks compat label?

I think that should be here just to confirm the fact.

File: modules/visual_script/visual_script.cpp (hunk 6 above)
Comment (jak6jak) [reply modules/visual_script/visual_script.cpp:585]: When editing a member, visual_script_editor.cpp _member_edited checks to make sure that the name is not already a function. variable, or signal. I assume that this reassuring that this still remains the case.

**Comment by akien-mga** [comment]
@swarnimarun Do you think you'll be able to address the review comments in the near future? It would be nice to get this merged so we can move on to the submodules PR.

**Comment by swarnimarun** [comment]
Aah sorry, I somehow got even more work in the last couple of weeks when I was supposed to be free so it got delayed, I ended up never having the time to really test stuff, I think I pushed all the formatting and such changes now though...

Will probably have to be fixing other bugs(if and when I find more) with further PRs...

**Comment by akien-mga** [comment]
Needs another rebase (sorry!), but aside from that I think this could likely be merged now, so that we can move to the next step?

WDYT @fire?

**Comment by fire** [comment]
Go go go!

On Nov 20, 2020, at 5:54 AM, Rémi Verschelde <notifications@github.com> wrote:

﻿
Needs another rebase (sorry!), but aside from that I think this could likely be merged now, so that we can move to the next step?

WDYT @fire?

**Comment by reduz** [comment]
I am unfamiliar with VS, so if @fire is ok, we should merge

**Comment by akien-mga** [comment]
Needs yet another rebase /o\
Ping me when it's done so I can merge it fast.

**Comment by swarnimarun** [comment]
Sorry been really busy during the finals, tbh, I missed the last message for the rebase( LOL ).. doing it asap.

**Comment by swarnimarun** [comment]
Have confirmed there doesn't seem to be anything that is broken after the rebase and changes as far as I can tell... so should be fine to merge.

Aside:
That iOS build fail was, as I understand, some kind of anomaly I didn't have to change anything, and on trying on a custom ios sandbox it compiles fine too... Even the logs referred to something not relevant to the code and though I did go through most of my code again on all compilers/platforms just to make sure...

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
>  That iOS build fail was, as I understand, some kind of anomaly I didn't have to change anything, and on trying on a custom ios sandbox it compiles fine too... Even the logs referred to something not relevant to the code and though I did go through most of my code again on all compilers/platforms just to make sure...

Yeah that happens every now and then, CI builds can sometime fail due to out of sync cache entries, or networking issues on the CI runners. When that happens restarting the failed build usually fixes it.

**Comment by akien-mga** [comment]
Thanks a ton! Looking forward to the next PR :)

**Comment by akien-mga** [comment]
There are some documentation changes needed, here's the output of `godot --doctool .`:
```diff
diff --git a/modules/visual_script/doc_classes/VisualScript.xml b/modules/visual_script/doc_classes/VisualScript.xml
index 088d84d2ec..c1c33056d0 100644
--- a/modules/visual_script/doc_classes/VisualScript.xml
+++ b/modules/visual_script/doc_classes/VisualScript.xml
@@ -26,23 +26,21 @@
 			</return>
 			<argument index="0" name="name" type="StringName">
 			</argument>
+			<argument index="1" name="arg1" type="int">
+			</argument>
 			<description>
-				Add a function with the specified name to the VisualScript.
 			</description>
 		</method>
 		<method name="add_node">
 			<return type="void">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="id" type="int">
+			<argument index="0" name="id" type="int">
 			</argument>
-			<argument index="2" name="node" type="VisualScriptNode">
+			<argument index="1" name="node" type="VisualScriptNode">
 			</argument>
-			<argument index="3" name="position" type="Vector2" default="Vector2( 0, 0 )">
+			<argument index="2" name="position" type="Vector2" default="Vector2( 0, 0 )">
 			</argument>
 			<description>
-				Add a node to a function of the VisualScript.
 			</description>
 		</method>
 		<method name="add_variable">
@@ -157,35 +155,29 @@
 		<method name="data_connect">
 			<return type="void">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="from_node" type="int">
+			<argument index="0" name="from_node" type="int">
 			</argument>
-			<argument index="2" name="from_port" type="int">
+			<argument index="1" name="from_port" type="int">
 			</argument>
-			<argument index="3" name="to_node" type="int">
+			<argument index="2" name="to_node" type="int">
 			</argument>
-			<argument index="4" name="to_port" type="int">
+			<argument index="3" name="to_port" type="int">
 			</argument>
 			<description>
-				Connect two data ports. The value of [code]from_node[/code]'s [code]from_port[/code] would be fed into [code]to_node[/code]'s [code]to_port[/code].
 			</description>
 		</method>
 		<method name="data_disconnect">
 			<return type="void">
 			</return>
-			<argument index="0" name="func" type="StringName">
+			<argument index="0" name="from_node" type="int">
 			</argument>
-			<argument index="1" name="from_node" type="int">
+			<argument index="1" name="from_port" type="int">
 			</argument>
-			<argument index="2" name="from_port" type="int">
+			<argument index="2" name="to_node" type="int">
 			</argument>
-			<argument index="3" name="to_node" type="int">
-			</argument>
-			<argument index="4" name="to_port" type="int">
+			<argument index="3" name="to_port" type="int">
 			</argument>
 			<description>
-				Disconnect two data ports previously connected with [method data_connect].
 			</description>
 		</method>
 		<method name="get_function_node_id" qualifiers="const">
@@ -197,35 +189,26 @@
 				Returns the id of a function's entry point node.
 			</description>
 		</method>
-		<method name="get_function_scroll" qualifiers="const">
-			<return type="Vector2">
+		<method name="get_node" qualifiers="const">
+			<return type="VisualScriptNode">
 			</return>
-			<argument index="0" name="name" type="StringName">
+			<argument index="0" name="id" type="int">
 			</argument>
 			<description>
-				Returns the position of the center of the screen for a given function.
 			</description>
 		</method>
-		<method name="get_node" qualifiers="const">
-			<return type="VisualScriptNode">
+		<method name="get_node_position" qualifiers="const">
+			<return type="Vector2">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="id" type="int">
+			<argument index="0" name="id" type="int">
 			</argument>
 			<description>
-				Returns a node given its id and its function.
 			</description>
 		</method>
-		<method name="get_node_position" qualifiers="const">
+		<method name="get_scroll" qualifiers="const">
 			<return type="Vector2">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="id" type="int">
-			</argument>
 			<description>
-				Returns a node's position in pixels.
 			</description>
 		</method>
 		<method name="get_variable_default_value" qualifiers="const">
@@ -267,18 +250,15 @@
 		<method name="has_data_connection" qualifiers="const">
 			<return type="bool">
 			</return>
-			<argument index="0" name="func" type="StringName">
+			<argument index="0" name="from_node" type="int">
 			</argument>
-			<argument index="1" name="from_node" type="int">
+			<argument index="1" name="from_port" type="int">
 			</argument>
-			<argument index="2" name="from_port" type="int">
+			<argument index="2" name="to_node" type="int">
 			</argument>
-			<argument index="3" name="to_node" type="int">
-			</argument>
-			<argument index="4" name="to_port" type="int">
+			<argument index="3" name="to_port" type="int">
 			</argument>
 			<description>
-				Returns whether the specified data ports are connected.
 			</description>
 		</method>
 		<method name="has_function" qualifiers="const">
@@ -293,27 +273,21 @@
 		<method name="has_node" qualifiers="const">
 			<return type="bool">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="id" type="int">
+			<argument index="0" name="id" type="int">
 			</argument>
 			<description>
-				Returns whether a node exists with the given id.
 			</description>
 		</method>
 		<method name="has_sequence_connection" qualifiers="const">
 			<return type="bool">
 			</return>
-			<argument index="0" name="func" type="StringName">
+			<argument index="0" name="from_node" type="int">
 			</argument>
-			<argument index="1" name="from_node" type="int">
+			<argument index="1" name="from_output" type="int">
 			</argument>
-			<argument index="2" name="from_output" type="int">
-			</argument>
-			<argument index="3" name="to_node" type="int">
+			<argument index="2" name="to_node" type="int">
 			</argument>
 			<description>
-				Returns whether the specified sequence ports are connected.
 			</description>
 		</method>
 		<method name="has_variable" qualifiers="const">
@@ -346,12 +320,9 @@
 		<method name="remove_node">
 			<return type="void">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="id" type="int">
+			<argument index="0" name="id" type="int">
 			</argument>
 			<description>
-				Remove a specific node.
 			</description>
 		</method>
 		<method name="remove_variable">
@@ -399,43 +370,25 @@
 		<method name="sequence_connect">
 			<return type="void">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="from_node" type="int">
+			<argument index="0" name="from_node" type="int">
 			</argument>
-			<argument index="2" name="from_output" type="int">
+			<argument index="1" name="from_output" type="int">
 			</argument>
-			<argument index="3" name="to_node" type="int">
+			<argument index="2" name="to_node" type="int">
 			</argument>
 			<description>
-				Connect two sequence ports. The execution will flow from of [code]from_node[/code]'s [code]from_output[/code] into [code]to_node[/code].
-				Unlike [method data_connect], there isn't a [code]to_port[/code], since the target node can have only one sequence port.
 			</description>
 		</method>
 		<method name="sequence_disconnect">
 			<return type="void">
 			</return>
-			<argument index="0" name="func" type="StringName">
-			</argument>
-			<argument index="1" name="from_node" type="int">
-			</argument>
-			<argument index="2" name="from_output" type="int">
+			<argument index="0" name="from_node" type="int">
 			</argument>
-			<argument index="3" name="to_node" type="int">
-			</argument>
-			<description>
-				Disconnect two sequence ports previously connected with [method sequence_connect].
-			</description>
-		</method>
-		<method name="set_function_scroll">
-			<return type="void">
-			</return>
-			<argument index="0" name="name" type="StringName">
+			<argument index="1" name="from_output" type="int">
 			</argument>
-			<argument index="1" name="ofs" type="Vector2">
+			<argument index="2" name="to_node" type="int">
 			</argument>
 			<description>
-				Position the center of the screen for a function.
 			</description>
 		</method>
 		<method name="set_instance_base_type">
@@ -450,14 +403,19 @@
 		<method name="set_node_position">
 			<return type="void">
 			</return>
-			<argument index="0" name="func" type="StringName">
+			<argument index="0" name="id" type="int">
 			</argument>
-			<argument index="1" name="id" type="int">
+			<argument index="1" name="position" type="Vector2">
 			</argument>
-			<argument index="2" name="position" type="Vector2">
+			<description>
+			</description>
+		</method>
+		<method name="set_scroll">
+			<return type="void">
+			</return>
+			<argument index="0" name="ofs" type="Vector2">
 			</argument>
 			<description>
-				Position a node on the screen.
 			</description>
 		</method>
 		<method name="set_variable_default_value">
@@ -496,9 +454,7 @@
 	</methods>
 	<signals>
 		<signal name="node_ports_changed">
-			<argument index="0" name="function" type="String">
-			</argument>
-			<argument index="1" name="id" type="int">
+			<argument index="0" name="id" type="int">
 			</argument>
 			<description>
 				Emitted when the ports of a node are changed.

```

**Comment by swarnimarun** [comment]
wooops fixing it... :P

**Pull request merged**